executor/explain.go

```diff
@@ -0,0 +1,216 @@
+	rows         int64
+	extra        []string
+}
+
+func (e *explainEntry) setJoinTypeForTableScan(p *plan.TableScan) {
+	if len(p.AccessConditions) == 0 {
+		e.joinType = "ALL"
+		return
+	}
+	if p.RefAccess {
+		e.joinType = "eq_ref"
+		return
+	}
+	for _, con := range p.AccessConditions {
+		if x, ok := con.(*ast.BinaryOperationExpr); ok {
+			if x.Op == opcode.EQ {
+				e.joinType = "const"
+				return
+			}
+		}
+	}
+	e.joinType = "range"
+}
+
+func (e *explainEntry) setJoinTypeForIndexScan(p *plan.IndexScan) {
+	if len(p.AccessConditions) == 0 {
+		e.joinType = "index"
+		return
+	}
+	if len(p.AccessConditions) == p.AccessEqualCount {
+		if p.RefAccess {
+			if p.Index.Unique {
+				e.joinType = "eq_ref"
+			} else {
+				e.joinType = "ref"
+			}
+		} else {
+			if p.Index.Unique {
+				e.joinType = "const"
+			} else {
+				e.joinType = "range"
+			}
+		}
+		return
+	}
+	e.joinType = "range"
+}
+
+// ExplainExec represents an explain executor.
+// See: https://dev.mysql.com/doc/refman/5.7/en/explain-output.html
+type ExplainExec struct {
+	StmtPlan plan.Plan
+	fields   []*ast.ResultField
+	rows     []*Row
+	cursor   int
+}
+
+// Fields implements Executor Fields interface.
+func (e *ExplainExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Execution Next interface.
+func (e *ExplainExec) Next() (*Row, error) {
+	if e.rows == nil {
+		e.fetchRows()
+	}
+	if e.cursor >= len(e.rows) {
+		return nil, nil
+	}
+	row := e.rows[e.cursor]
+	e.cursor++
+	return row, nil
+}
+
+func (e *ExplainExec) fetchRows() {
+	visitor := &explainVisitor{id: 1}
+	e.StmtPlan.Accept(visitor)
+	for _, entry := range visitor.entries {
+		row := &Row{}
+		row.Data = []interface{}{
+			entry.ID,
+			entry.selectType,
+			entry.table,
+			entry.joinType,
+			entry.key,
+			entry.key,
+			entry.keyLen,
+			entry.ref,
+			entry.rows,
+			strings.Join(entry.extra, "; "),
+		}
+		for i := range row.Data {
+			if row.Data[i] == "" {
+				row.Data[i] = nil
+			}
+		}
+		e.rows = append(e.rows, row)
+	}
+}
+
+// Close implements Executor Close interface.
+func (e *ExplainExec) Close() error {
+	return nil
+}
+
+type explainVisitor struct {
+	id int64
+
+	// Sort extra should be appended in the first table in a join.
+	sort    bool
+	entries []*explainEntry
+}
+
+func (v *explainVisitor) Enter(p plan.Plan) (plan.Plan, bool) {
+	switch x := p.(type) {
+	case *plan.TableScan:
+		v.entries = append(v.entries, v.newEntryForTableScan(x))
+	case *plan.IndexScan:
+		v.entries = append(v.entries, v.newEntryForIndexScan(x))
+	case *plan.Sort:
+		v.sort = true
+	}
+	return p, false
+}
+
+func (v *explainVisitor) Leave(p plan.Plan) (plan.Plan, bool) {
+	return p, true
+}
+
+func (v *explainVisitor) newEntryForTableScan(p *plan.TableScan) *explainEntry {
+	entry := &explainEntry{
+		ID:         v.id,
+		selectType: "SIMPLE",
+		table:      p.Table.Name.O,
+	}
+	entry.setJoinTypeForTableScan(p)
+	if entry.joinType != "ALL" {
+		entry.key = "PRIMARY"
```

Comment:
join type “ALL" means all rows will be scanned. otherwise, the primary key handle must used.